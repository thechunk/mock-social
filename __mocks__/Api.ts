module.exports = {
    UsersApi: {
        getAll: () => Promise.resolve([])
    }
};