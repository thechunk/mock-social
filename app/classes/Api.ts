import Config from 'react-native-config';

enum ApiPath {
    Users = '/users',
    Albums = '/albums',
    Photos = '/photos'
}
enum ApiQueryParam {
    UserId = 'userId',
    AlbumId = 'albumId'
}
enum HttpMethod {
    Get = 'GET',
    Post = 'POST',
    Put = 'PUT',
    Patch = 'PATCH',
    Delete = 'DELETE'
}
enum MimeType {
    ApplicationJson = 'application/json'
}

const apiUrl = (baseUrl: string, path: ApiPath, query?: string): RequestInfo => {
    return `${baseUrl}${path}${query ? `?${query}` : ''}` as RequestInfo;
};

const apiRequest = <T extends any>(url: RequestInfo, method?: HttpMethod, body?: string) => ((): Promise<T> => (
    fetch(url, {
        method: method ? method : HttpMethod.Get,
        headers: {
            Accept: MimeType.ApplicationJson,
            'Content-Type': MimeType.ApplicationJson
        },
        body: body != null && body.length > 0 ? JSON.stringify(body) : null
    })
        .then((response): T => response.json() as any as T)
));
const apiBaseUrl = Config.API_BASE_URL;

class UsersApi implements IUsersApi {
    public getAll() {
        return apiRequest<Array<IUser>>(apiUrl(apiBaseUrl, ApiPath.Users))()
    }

    public getById(id: number) {
        return apiRequest<IUser>(`${apiUrl(apiBaseUrl, ApiPath.Users)}/${id}`)()
    }
}
const Users = new UsersApi();

class AlbumsApi implements IAlbumsApi {
    public getAllByUserId(id: number) {
        const query = `${ApiQueryParam.UserId}=${id}`;
        return apiRequest<Array<IAlbum>>(apiUrl(apiBaseUrl, ApiPath.Albums, query))()
    }
}
const Albums = new AlbumsApi();

class PhotosApi implements IPhotosApi {
    public getAllByAlbumId(id: number) {
        const query = `${ApiQueryParam.AlbumId}=${id}`;
        return apiRequest<Array<IPhoto>>(apiUrl(apiBaseUrl, ApiPath.Photos, query))()
    }
}
const Photos = new PhotosApi();

export {Users as UsersApi};
export {Albums as AlbumsApi};
export {Photos as PhotosApi};
