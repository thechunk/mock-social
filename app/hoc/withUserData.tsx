import React, {ComponentType, FC, useEffect, useState} from 'react';
import {UsersApi} from '../classes/Api';

const withUserData = <P extends IWithUserDataProps>(Component: ComponentType<P>):
    FC<HocOptional<P, IWithUserDataProps> & IWithUserDataOptionalProps> =>
    (props: HocOptional<P, IWithUserDataProps> & IWithUserDataOptionalProps) => {
        const [user, setUser] = useState<IUser | null>(null);
        useEffect(() => {
            if (props.id) UsersApi.getById(props.id).then(setUser);
        }, [props.id]);

        useEffect(() => {
            if (props.hydrate) setUser(props.hydrate);
        }, [props.hydrate]);

        return (
            <Component {...props as P & IWithUserDataOptionalProps} user={user} />
        );
    };

export default withUserData;
