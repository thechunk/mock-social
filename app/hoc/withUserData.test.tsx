/**
 * @jest-environment jsdom
 */

import * as React from 'react';
import {mount, shallow} from 'enzyme';
import {View} from "react-native";
import withUserData from "./withUserData";
import {UsersApi} from '../classes/Api';
require('setimmediate');

describe('withUserData', () => {
    it('wrapper renders correctly', () => {
        const Component = () => (<View/>);
        const Wrapped: any = withUserData(Component);
        expect(shallow(<Wrapped id={1} />)).toMatchSnapshot();
    });

    it('hydrates with initial data', () => {
        const Component = () => (<div />);
        const Wrapped: any = withUserData(Component);
        const wrappedComponent = mount(<Wrapped hydrate={{foo: 'bar'}} />).childAt(0);
        expect(wrappedComponent.props().user).toEqual({foo: 'bar'});
    });

    it('calls api', (done) => {
        UsersApi.getById = jest.fn().mockReturnValue(Promise.resolve());
        const Component = () => (<div />);
        const Wrapped: any = withUserData(Component);
        mount(<Wrapped id={1} />);
        done();
        expect(UsersApi.getById).toBeCalledTimes(1);
    });

    it('calls api, sets data and loader status on success', (done) => {
        UsersApi.getById = jest.fn().mockReturnValue(Promise.resolve({bar: 'foo'}));
        const Component = () => (<div />);
        const Wrapped: any = withUserData(Component);
        let wrappedComponent = mount(<Wrapped id={1} />).childAt(0);
        expect(wrappedComponent.props().loading).toEqual(true);

        flushAwaitable(done).then(() => {
            expect(wrappedComponent.props().loading).toEqual(false);
            expect(wrappedComponent.props().user).toEqual({bar: 'foo'});
        });
    });
});
