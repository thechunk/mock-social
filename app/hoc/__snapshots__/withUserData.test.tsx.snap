// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUserData wrapper renders correctly 1`] = `
<Component
  id={1}
  loading={true}
  user={null}
/>
`;
